source/screens/FlashUtils.cpp

```diff
@@ -0,0 +1,217 @@
+#include "FlashUtils.hpp"
+#include <cstdio>
+
+#include <coreinit/mcp.h>
+#include <coreinit/thread.h>
+#include <coreinit/filesystem_fsa.h>
+#include <nsysccr/cdc.h>
+#include <nsysccr/cfg.h>
+#include <nn/ccr.h>
+
+FlashUtils flashUtils;
+
+bool FlashUtils::CopyFile(const std::string& srcPath, const std::string& dstPath)
+{
+    FILE* inf = fopen(srcPath.c_str(), "rb");
+    if (!inf) {
+        return false;
+    }
+
+    FILE* outf = fopen(dstPath.c_str(), "wb");
+    if (!outf) {
+        fclose(inf);
+        return false;
+    }
+
+    uint8_t buf[4096];
+    size_t bytesRead;
+    while ((bytesRead = fread(buf, 1, sizeof(buf), inf)) > 0) {
+        if (fwrite(buf, 1, bytesRead, outf) != bytesRead) {
+            fclose(inf);
+            fclose(outf);
+            return false;
+        }
+    }
+
+    if (ferror(inf)) {
+        fclose(inf);
+        fclose(outf);
+        return false;
+    }
+
+    fclose(inf);
+    fclose(outf);
+    return true;
+}
+
+bool FlashUtils::ReadFirmwareHeader(const std::string& path, FlashUtils::FirmwareHeader& header)
+{
+    FILE* f = fopen(path.c_str(), "rb");
+    if (!f) {
+        return false;
+    }
+
+    if (fread(&header, 1, sizeof(header), f) != sizeof(header)) {
+        fclose(f);
+        return false;
+    }
+
+    fclose(f);
+    return true;
+}
+
+bool FlashUtils::CheckVersionSafety(const uint32_t firmver, const uint32_t langver)
+{
+    // Very crude way to do the checks universally, but it works well enough
+    // We take the firmware version as well as the language version (w/o region - am I stupid?) and compare them 
+    if ((firmver == 0x190c0117 || firmver == 0xfe000000) && ((langver & 0xFFFF00) ^ 0x170200)==0){ // Only patched firmware uses language v5890
+        return true;
+    } else if (firmver == 0x18140116 && not ((langver & 0xFFFF00) ^ 0x160400)){
+        return true;
+    } else if (firmver == 0x17080114 && not ((langver & 0xFFFF00) ^ 0x140000)){
+        return true;
+    } else if ((firmver == 0x151e0113 || firmver == 0x15060113 || firmver == 0x14060113) && not ((langver & 0xFFFF00) ^ 0x130000)){ // 3 versions have used language v4864
+        return true;
+    } else {
+        return false;
+    }
+}
```

Comment:
Not sure how comfortable I feel about this. Also why `not ((langver & 0xFFFF00) ^ 0x160400)` instead of just `(langver & 0xFFFF00) == 0x160400)`?

Reply:
This works right now to detect version (mis)matches and whenever we'll need an expansion, one will be added. That will probably never happen, though.
I've checked and seemingly all cases in here pass this check. This function shouldn't even be called in the case that we can't retrieve a value to check on.

Comparisons also simplified.